tests/cover/test_type_lookup.py

```diff
@@ -0,0 +1,140 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import pytest
+
+import hypothesis.strategies as st
+from hypothesis import given, infer
+from hypothesis.errors import InvalidArgument, ResolutionFailed
+from hypothesis.strategies import from_type
+from hypothesis.searchstrategy import types
+from hypothesis.internal.compat import PY2, integer_types
+
+# Build a set of all types output by core strategies
+blacklist = [
+    'builds', 'iterables', 'permutations', 'random_module', 'randoms',
+    'runner', 'sampled_from', 'streaming',
+]
+types_with_core_strat = set(integer_types)
+for thing in (getattr(st, name) for name in sorted(st._strategies)
+              if name in dir(st) and name not in blacklist):
+    for n in range(3):
+        try:
+            ex = thing(*([st.nothing()] * n)).example()
+            types_with_core_strat.add(type(ex))
+            break
+        except (TypeError, InvalidArgument):
+            continue
+
+
+@pytest.mark.parametrize('typ', sorted(types_with_core_strat, key=str))
+def test_resolve_core_strategies(typ):
+    @given(from_type(typ))
+    def inner(ex):
+        if PY2 and issubclass(typ, integer_types):
+            assert isinstance(ex, integer_types)
+        else:
+            assert isinstance(ex, typ)
+
+    inner()
+
+
+def test_lookup_knows_about_all_core_strategies():
+    cannot_lookup = types_with_core_strat - set(types.type_strategy_mapping())
+    assert not cannot_lookup
+
+
+def test_lookup_keys_are_types():
+    with pytest.raises(InvalidArgument):
+        st.register_type_strategy('int', st.integers())
+    assert 'int' not in types._global_type_to_strategy_lookup
+
+
+def test_lookup_values_are_strategies():
+    with pytest.raises(InvalidArgument):
+        st.register_type_strategy(int, 42)
+    assert 42 not in types._global_type_to_strategy_lookup.values()
+
+
+@pytest.mark.parametrize('typ', sorted(types_with_core_strat, key=str))
+def test_lookup_overrides_defaults(typ):
+    sentinel = object()
+    try:
+        st.register_type_strategy(typ, st.just(sentinel))
+        assert from_type(typ).example() is sentinel
+    finally:
```

Comment:
I feel like these tests might work better using a `teardown_function` or a fixture to reset the whole mapping after every test rather than trying to get it right in each test.